### ISSUE TITLE
GPU usage low

### ISSUE THREAD

**Issue by Xunzhaocunzi** [pierluigiferrari/ssd_keras#109]
Hi @pierluigiferrari

Your tutorials are really amazing, they help me a lot in understanding SSD architecture. I appreciated. 

In addition, I am following your tutorial to repeat ssd7 model training process, and I found during training GPU usage was really low( about 5% with my gtx 1070 laptop version) when batch size =32,  I guess it might be caused by data generator. Then, I increased batch size=64, however, it reported out of memory issue. 

Do you have any suggestions regarding training? 

Thank you
![oom](https://user-images.githubusercontent.com/32579634/39226586-e968397a-4818-11e8-96d3-35d5032c25d9.PNG)

**Comment by Xunzhaocunzi** [comment]
Hello @pierluigiferrari 

I tried again on my lab's desktop (2 GTX1080Ti, and 64gb memory), and it generated the same result; crashed when batchsize =64. So I guess probably there is something wrong with the "data generator" codes. How can I train the model without generator part? Thank you

**Comment by pierluigiferrari** [comment]
Data augmentation and loading compressed images (JPG) is simply computationally expensive and if you have a very small model like SSD7, the generator becomes the bottleneck. It's less of an issue the more expensive the forward and backward pass of the model becomes, i.e. for more complex models.

Two ways to speed up the data generation would be to have the image data uncompressed and contiguous, e.g. in an HDF5 file or TFRecord, and to parallelize the data generation, i.e. to use multiprocessing to generate batches on multiple CPU cores. `DataGenerator` currently does neither of these two things.

You can always replace the provided data generator with something faster. Let me know in case you find a good alternative.

**Comment by pierluigiferrari** [comment]
`DataGenerator` is now able to convert a dataset into an HDF5 file that stores all images as uncompressed arrays in a contiguous block of memory, which speeds up the data generation a bit.

I'll close this issue since I consider your question answered, but feel free to reopen if that's not the case.

**Comment by Xunzhaocunzi** [comment]
Thank you very much. I will try to train in AWS to see if it can speed up.

**Comment by Xunzhaocunzi** [comment]
I ran the ssd7 model on AWS with g3-4large GPU, and the speed seems slow too. For one single epoch, it may take 65 mins to complete. 

How did the original caffe model do for data augmentation? 

Anyway, I appreciated that you could share this great keras model with us.   

![capture](https://user-images.githubusercontent.com/32579634/39402735-c49310d0-4b2d-11e8-9e88-9bd6c3015f9e.PNG)

**Comment by pierluigiferrari** [comment]
I cannot confirm these numbers. Running 1000 training steps with the presets in the SSD7 training notebook takes around 13 minutes on my laptop (Ubuntu 16.04, Intel Core i7-7700HQ quad core, GeForce GTX 1070). This is when serving the dataset from an HDF5 file.

> How did the original caffe model do for data augmentation?

According to various issues in the GitHub repository of the original Caffe implementation, one full training (120,000 steps) of the SSD300 "07+12" model takes roughly 60 hours on one TitanX GPU, which means around 30 minutes per 1000 training steps. This is in line with what I'm getting for this Keras implementation. My training also took around 60 hours in total.

**Comment by Xunzhaocunzi** [comment]
Thank you for your reply. 

I am using CUDA 9.0, CuDNN 7.0 and windows 10. 
When I started training, I got this warning "Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX AVX2". Maybe I need make sure my CPU to use AVX and AVX2 at data generator? 

My desktop has i7 8700K and 1080Ti SLI, thus I was quite frustrated that it took so long for one single epoch -.-
